plugins/org.polarsys.capella.scenario.editor.dsl/src/org/polarsys/capella/scenario/editor/dsl/validation/TextualScenarioValidator.xtend

```diff
@@ -152,24 +153,40 @@ class TextualScenarioValidator extends AbstractTextualScenarioValidator {
 			error(String.format("Insert timeline"), TextualScenarioPackage.Literals.STATE_FRAGMENT__TIMELINE)
 			return
 		}
+
+		if (!EmbeddedEditorInstanceHelper.checkValidTimeline(fragment.timeline)) {
+			error(String.format("Timeline is not present in diagram", fragment.keyword),
+				TextualScenarioPackage.Literals.STATE_FRAGMENT__TIMELINE)
+			return
+		}
+
 		if (fragment.keyword === null) {
-			error(String.format("Insert \"state\", \"mode\" or \"function\""), TextualScenarioPackage.Literals.STATE_FRAGMENT__KEYWORD)
+			error(String.format("Insert \"state\", \"mode\" or \"function\""),
+				TextualScenarioPackage.Literals.STATE_FRAGMENT__KEYWORD)
 			return
 		}
-		
-		if (fragment.name === null ) {
-			error(String.format("Insert the " + fragment.keyword + " name"), TextualScenarioPackage.Literals.STATE_FRAGMENT__NAME)
+
+		var scenarioType = EmbeddedEditorInstanceHelper.getScenarioType();
+		if (fragment.keyword.equals(DslConstants.FUNCTION) && scenarioType.equals(DslConstants.FUNCTIONAL)) {
+			error(String.format("\'function\' can not be used in this diagram"),
+				TextualScenarioPackage.Literals.STATE_FRAGMENT__KEYWORD)
 			return
 		}
-		
-		var availableStateFragments = EmbeddedEditorInstanceHelper.getAvailableStateFragments(fragment.keyword, fragment.timeline)
-		if (!availableStateFragments.contains(fragment.name) ) {
-			error(String.format("This " + fragment.keyword + " does not exist or it's not available for " + fragment.timeline), TextualScenarioPackage.Literals.STATE_FRAGMENT__NAME)
+
+		if (fragment.name === null) {
+			error(String.format("Insert the " + fragment.keyword + " name"),
+				TextualScenarioPackage.Literals.STATE_FRAGMENT__NAME)
+			return
 		}
-		
-		if (!EmbeddedEditorInstanceHelper.checkValidTimeline(fragment.timeline)){
-			error(String.format("This timeline does not exist", fragment.keyword), TextualScenarioPackage.Literals.STATE_FRAGMENT__TIMELINE)
+
+		var availableStateFragments = EmbeddedEditorInstanceHelper.getAvailableStateFragments(fragment.keyword,
+			fragment.timeline)
+		if (!availableStateFragments.contains(fragment.name)) {
+			error(
+				String.format("This " + fragment.keyword + " does not exist or it's not available for " +
+					fragment.timeline), TextualScenarioPackage.Literals.STATE_FRAGMENT__NAME)
 		}
+
 	}

```

Comment:
add ' ' to surround keyword:
"This '" + fragment.keyword + "' does not exist or it's not available for " +
					fragment.timeline)